### DISCUSSION TITLE
Question: How to send attachment messages using Chatwoot API

### DISCUSSION THREAD

**Discussion by NitinNR** [chatwoot/chatwoot#1809] [Q&A] [answered]
self hosted on linux

**Comment by hartois** [comment]
Use POST multipart/form-data request to /accounts/{account_id}/conversations/{id}/messages with attachments[] field

**Reply by NitinNR** [reply]
i need to send both images and files what exactly the format will be?

**Comment by pranavrajs** [comment]
https://github.com/chatwoot/chatwoot/issues/1636#issuecomment-757514732

@NitinNR Hope this link helps.

**Comment by NitinNR** [comment]
what should i change in this json request for sending files and images?

dictToSend = {
      "content": message,
      "message_type": msg_type,
      "private": False,
      "content_type": "cards",
      "content_attributes": {
        "items":[
          {"actions":[{"text":message }]}
        ]  
      }
    }

**Reply by hartois** [reply]
For attach files you need use multipart/form-data instead of JSON.
Example from my PHP application:
```
public function buildDataFiles($boundary, $fields, $files){
        $data = '';
        $eol = "\r\n";

        $delimiter = '-------------' . $boundary;

        foreach ($fields as $name => $content) {
            $data .= "--" . $delimiter . $eol
                . 'Content-Disposition: form-data; name="' . $name . "\"".$eol.$eol
                . $content . $eol;
        }


        foreach ($files as $name => $file) {
            $content = $file['content'];
            $filename = $file['filename'];
            $mime = (!empty($file['mime'])) ? 'Content-Type: '.$file['mime'] . $eol : '';
            $data .= "--" . $delimiter . $eol
                . 'Content-Disposition: form-data; name="' . $name . '"; filename="' . $filename.'"' . $eol
                . $mime
                . 'Content-Transfer-Encoding: binary'.$eol
            ;

            $data .= $eol;
            $data .= $content . $eol;
        }
        $data .= "--" . $delimiter . "--".$eol;


        return $data;
    }

    public function newAttachment(int $accountId, int $conversationId, ?string $text,
                                  string $path, string $fileName, string $messageType, bool $isPrivate){

        $content = file_get_contents($path);
        $finfo = new \finfo(FILEINFO_MIME_TYPE);
        $files = ['attachments[]'=>[
            'content'=>$content,
            'filename'=>$fileName,
            'mime'=>$finfo->buffer($content)
        ]];

        $bound = uniqid();
        $data = $this->buildDataFiles($bound, [
            'content'=>$text, 'message_type' => $messageType, 'private' => $isPrivate
        ], $files);

        $headers = ['Content-Type: multipart/form-data; boundary=-------------'.$bound,
            'Content-Length: ' . strlen($data)];

        $url = '/accounts/'.$accountId.'/conversations/'.$conversationId.'/messages';
        $this->setHeaders(['api_access_token: '.$this->cwCfg->get('apiKey')]);

        return $this->callApi($url,'POST', [], $data, $headers);
    }


```

**Reply by NitinNR** [reply]
please give me the json format i need to use in python
request

**Reply by NitinNR** [reply]
this is not also working please provide some support
{
   "content_type":"cards",
   "content_attributes":{
      "items":[
         {
            "media_url":"https://assets.ajio.com/medias/sys_master/root/hdb/h9a/13582024212510/-1117Wx1400H-460345219-white-MODEL.jpg",
            "title":"Nike Shoes 2.0",
            "description":"Running with Nike Shoe 2.0",
            "actions":[
               {
                  "type":"link",
                  "text":"View More",
                  "uri":"google.com"
               },
               {
                  "type":"postback",
                  "text":"Add to cart",
                  "payload":"ITEM_SELECTED"
               }
            ]
         }
      ]
   },
   "private":false
}

**Reply by hartois** [reply]
You cannot post file data using JSON format.
You must use multipart/form-data.
try read https://stackoverflow.com/questions/49487778/python-http-post-with-upload-file-and-headers-generated-from-postman

**Comment by sojan-official** [comment, marked as answer]
> POST multipart/form-data request to /accounts/{account_id}/conversations/{id}/messages with attachments[] field



## Creating an attachment message 

Postman link: https://www.postman.com/chatwoot/workspace/chatwoot-apis/request/162770-e22922ab-27e5-4d82-a672-a52f6555ed21

*headers*
<img width="876" alt="Screenshot 2021-08-24 at 8 48 00 PM" src="https://user-images.githubusercontent.com/73185/130643622-af4b111b-9488-435d-a88b-0a297c8c59b6.png">


*body*
<img width="1683" alt="Screenshot 2021-08-24 at 8 35 41 PM" src="https://user-images.githubusercontent.com/73185/130642797-b0735b66-1454-496c-b4df-a5b9ce07447d.png">

### curl
```
curl --location --request POST 'http://localhost:3000/api/v1/accounts/1/conversations/2/messages' \
--header 'Content-Type: multipart/form-data; boundary=----WebKitFormBoundary' \
--header 'api_access_token: uBJjA56hAuS6fWEt6V1XRei8' \
--form 'attachments[]=@"K6aITBLXf/chatwoot.png"'
```

### PHP 
```
<?php
require_once 'HTTP/Request2.php';
$request = new HTTP_Request2();
$request->setUrl('http://localhost:3000/api/v1/accounts/1/conversations/2/messages');
$request->setMethod(HTTP_Request2::METHOD_POST);
$request->setConfig(array(
  'follow_redirects' => TRUE
));
$request->setHeader(array(
  'Content-Type' => 'multipart/form-data; boundary=----WebKitFormBoundary',
  'api_access_token' => 'uBJjA56hAuS6fWEt6V1XRei8'
));
$request->addUpload('attachments[]', 'K6aITBLXf/chatwoot.png', '[PROXY]', '<Content-Type Header>');
try {
  $response = $request->send();
  if ($response->getStatus() == 200) {
    echo $response->getBody();
  }
  else {
    echo 'Unexpected HTTP status: ' . $response->getStatus() . ' ' .
    $response->getReasonPhrase();
  }
}
catch(HTTP_Request2_Exception $e) {
  echo 'Error: ' . $e->getMessage();
}
```

### ruby 

```
require "uri"
require "net/http"

url = URI("http://localhost:3000/api/v1/accounts/1/conversations/2/messages")

http = Net::HTTP.new(url.host, url.port);
request = Net::HTTP::Post.new(url)
request["Content-Type"] = "multipart/form-data; boundary=----WebKitFormBoundary"
request["api_access_token"] = "uBJjA56hAuS6fWEt6V1XRei8"
form_data = [['attachments[]', File.open('K6aITBLXf/chatwoot.png')]]
request.set_form form_data, 'multipart/form-data'
response = http.request(request)
puts response.read_body
```


After creating the attachment message, it becomes accessible via the conversation API
postman: https://www.postman.com/chatwoot/workspace/chatwoot-apis/request/162770-7100f77a-e9b1-4508-aa4f-8900cf9568b6

<img width="1674" alt="Screenshot 2021-08-24 at 8 40 09 PM" src="https://user-images.githubusercontent.com/73185/130643287-d13769d0-b70a-4ddc-a142-de15cc7dd26f.png">

**Reply by dorain47** [reply]
@sojan-official This works perfectly to send the image as outgoing message, but how can the API be used to send image as incoming message?

**Reply by sojan-official** [reply]
@dorain47  in the form data add the field `message_type` to be 'incoming'

**Comment by thundersparkf** [comment]
Is there anyway to send a url for the file instead of uploading the file itself for attachment? @sojan-official

**Reply by sojan-official** [reply]
not at the moment. But feel free to raise a GitHub issue over which we can track it as an enhancement. since we support file URLs for avatar uploads.

**Comment by silvaartur** [comment]
Hello @sojan-official!

I have a doubt: when I follow the instructions according to your Postman I got this...

<img width="248" alt="Captura de Tela 2022-05-13 às 08 01 30" src="https://user-images.githubusercontent.com/12688324/168270253-e9b41a77-a073-44cc-82a7-bc1a775272fb.png">

...but I need this (another lib using ChatWoot API):

<img width="386" alt="Captura de Tela 2022-05-13 às 08 02 58" src="https://user-images.githubusercontent.com/12688324/168270414-c9774275-e87a-44fd-92ea-5a7ba5d4ff85.png">

So I need to show the preview for images and audio files.

Can u help me pointing what I doing wrong?

Tks.

**Reply by sojan-official** [reply]
@silvaartur  Chatwoot identifies the file type based on the mime-type of the uploaded file and renders a preview for the supported file type. So there shouldn't be any extra steps. are both the uploads the same file type?
